### PR TITLE
fix(Scripts/Hyjal): Archimonde refactored

### PR DESCRIPTION
**Pull request by elthehablo**
## Changes Proposed:

This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [x] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:

- Closes 

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [ ] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [x] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1. complete the rest of Hyjal
2. engage Archimonde
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by pklloveyou** [comment]
I tried it and it felt good to me

**Comment by kissingers** [comment]
> I tried it and it felt good to me

Has any testing of the finger of death protection when the boss is below 10% blood, and there is more than 1 player? Will it kill anyone other than the current target?

**Comment by elthehablo** [comment]
I can test that tonight prob

File: src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/boss_archimonde.cpp [hunk 1]
```diff
@@ -88,707 +88,473 @@ enum Events
...
+    {
+        if (Map* map = me->GetMap())
         {
-            Talk(SAY_SLAY);
-
-            if (victim && victim->GetTypeId() == TYPEID_PLAYER)
-                GainSoulCharge(victim->ToPlayer());
+            map->DoForAllPlayers([&](Player* player)
+            {
+                player->AddAura(SPELL_PROTECTION_OF_ELUNE, player);
```
Comment (pklloveyou) [inline src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/boss_archimonde.cpp:320]:
```
                        if (me->GetVictim() != player)
                            continue;
```

File: src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/boss_archimonde.cpp [hunk 2]
```diff
@@ -88,707 +88,473 @@ enum Events
...
+    void DoCastProtection()
+    {
+        if (Map* map = me->GetMap())
         {
-            Talk(SAY_SLAY);
-
-            if (victim && victim->GetTypeId() == TYPEID_PLAYER)
-                GainSoulCharge(victim->ToPlayer());
+            map->DoForAllPlayers([&](Player* player)
+            {
```
Comment (pklloveyou) [inline src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/boss_archimonde.cpp:319]:
```
+ if (me->GetVictim() != player)
+     continue;
```

Comment (pklloveyou) [reply src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/boss_archimonde.cpp:319]: image

Comment (elthehablo) [reply src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/boss_archimonde.cpp:319]: Wdym doforallplayers only iterates over players by definition

Comment (pklloveyou) [reply src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/boss_archimonde.cpp:319]: Have you tested it?

Comment (elthehablo) [reply src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/boss_archimonde.cpp:319]: I don't understand what you're getting at.

image
GetPlayers() gets all players, and not other creatures

**Comment by elthehablo** [comment]
Tested this once again and all seems to be in order. Should be good to merge

File: src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/boss_archimonde.cpp (hunk 2 above)
Comment (pklloveyou) [reply src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/boss_archimonde.cpp:319]: OK..

**Review by pangolp — APPROVED** [review]
I want to clarify that I did not do the tests. I give my validation, knowing that you made them, so that in any case, you can merge it. But I don't do any testing on this. But I trust that you did.

**Pull request merged**